components/lib/jexl/src/main/java/mozilla/components/lib/jexl/value/JexlValue.kt

```diff
@@ -0,0 +1,304 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.lib.jexl.value
+
+import mozilla.components.lib.jexl.evaluator.EvaluatorException
+
+/**
+ * A JEXL value type.
+ */
+sealed class JexlValue {
+    abstract operator fun plus(other: JexlValue): JexlValue
+    abstract operator fun times(other: JexlValue): JexlValue
+    abstract operator fun div(other: JexlValue): JexlValue
+    abstract operator fun compareTo(other: JexlValue): Int
+
+    abstract fun toBoolean(): Boolean
+    abstract fun toKotlin(): Any
+}
+
+/**
+ * JEXL Integer type.
+ */
+class JexlInteger(val value: Int): JexlValue() {
+    override fun div(other: JexlValue): JexlValue {
+        return when (other) {
+            is JexlInteger -> JexlInteger(value / other.value)
+            is JexlDouble -> JexlDouble(value / other.value)
+            else -> throw EvaluatorException("Can't divide by ${other::class}")
+        }
+    }
+
+    override fun times(other: JexlValue): JexlValue {
+        return when (other) {
+            is JexlInteger -> JexlInteger(value * other.value)
+            is JexlDouble -> JexlDouble(value * other.value)
+            is JexlBoolean -> JexlInteger(value * other.toInt())
+            else -> throw EvaluatorException("Can't multiply with ${other::class}")
+        }
+    }
+
+    override fun plus(other: JexlValue): JexlValue {
+        return when (other) {
+            is JexlInteger -> JexlInteger(value + other.value)
+            is JexlDouble -> JexlDouble(value + other.value)
+            is JexlString -> JexlString(value.toString() + other.value)
+            is JexlBoolean -> JexlInteger(value + (other.toInt()))
+            else -> throw EvaluatorException("Can't add ${other::class}")
+        }
+    }
+
+    override fun compareTo(other: JexlValue): Int {
+        return when (other) {
+            is JexlInteger -> value.compareTo(other.value)
+            is JexlDouble -> value.compareTo(other.value)
+            else -> throw EvaluatorException("Can't compare ${other::class}")
+        }
+    }
+
+    override fun equals(other: Any?) = other is JexlInteger && value == other.value
+
+    override fun toBoolean(): Boolean = value != 0
+
+    override fun toKotlin() = value
+
+    override fun toString() = value.toString()
+
+    override fun hashCode() = value.hashCode()
+}
+
+/**
+ * JEXL Double type.
+ */
+class JexlDouble(val value: Double): JexlValue() {
```

Comment:
Looks like a `JexlNumber` base class would make sense to reduce the duplication in `div/plus/minus/times`.

Reply:
This would also be an easy to way to get coverage up. `JexlDouble` has no coverage right now and then would have quite a bit. ;)

Reply:
Yeah, I plan to raise the coverage before landing!

The `Number` refactoring is something I would like to file a follow-up for.